### PR TITLE
[BEAM-7828] Fixes Key type conversion from/to client entity in Python Datastore IO

### PR DESCRIPTION
**Pull request by post-l**
* Currently, `ReadFromDatastore` doesn't convert the properties of type key so the user has to deal with `google.cloud.datastore.key.Key` type instead of the expected `apache_beam.io.gcp.datastore.v1new.datastoreio.Key` type. This is also the case when trying to write a key property using `WriteToDatastore`.
* `WriteToDatastore` and `DeleteFromDatastore` have an unused project ID argument. Like the previous version of the library, user expects the keys project attribute to be inferred from this parameter if unset.
* When comparing keys, user expects the keys to be equal only if they are both in the same project.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by post-l** [comment]
R: @udim

File: sdks/python/apache_beam/io/gcp/datastore/v1new/datastoreio.py [hunk 1]
```diff
@@ -382,6 +382,8 @@ def add_element_to_batch(self, element):
       if not isinstance(element, types.Entity):
         raise ValueError('apache_beam.io.gcp.datastore.v1new.datastoreio.Entity'
                          ' expected, got: %s' % type(element))
+      if not element.key.project:
```
Comment (udim) [inline sdks/python/apache_beam/io/gcp/datastore/v1new/datastoreio.py]: I can't comment there, but could you add to the comment on line 368 that missing project values will be filled with the transform's value?

File: sdks/python/apache_beam/io/gcp/datastore/v1new/datastoreio.py [hunk 2]
```diff
@@ -420,6 +422,8 @@ def add_element_to_batch(self, element):
       if not isinstance(element, types.Key):
         raise ValueError('apache_beam.io.gcp.datastore.v1new.datastoreio.Key'
                          ' expected, got: %s' % type(element))
+      if not element.project:
```
Comment (udim) [inline sdks/python/apache_beam/io/gcp/datastore/v1new/datastoreio.py]: Same as above.

File: sdks/python/apache_beam/io/gcp/datastore/v1new/types_test.py [hunk 3]
```diff
@@ -51,18 +51,22 @@ def testEntityToClientEntity(self):
...
     e = Entity(k, exclude_from_indexes=exclude_from_indexes)
-    e.set_properties({'efi1': 'value', 'property': 'value'})
+    ref = Key(['kind2', 1235])
+    e.set_properties({'efi1': 'value', 'property': 'value', 'ref': ref})
     ec = e.to_client_entity()
     self.assertEqual(kc, ec.key)
     self.assertSetEqual(set(exclude_from_indexes), ec.exclude_from_indexes)
     self.assertEqual('kind', ec.kind)
     self.assertEqual(1234, ec.id)
+    self.assertEqual(1235, ec['ref'].id)
```
[the remark this answers is not in the document]
Comment (post-l) [reply sdks/python/apache_beam/io/gcp/datastore/v1new/types_test.py]: done

File: sdks/python/apache_beam/io/gcp/datastore/v1new/datastoreio.py (hunk 1 above)
Comment (post-l) [reply sdks/python/apache_beam/io/gcp/datastore/v1new/datastoreio.py]: done

File: sdks/python/apache_beam/io/gcp/datastore/v1new/datastoreio.py (hunk 2 above)
Comment (post-l) [reply sdks/python/apache_beam/io/gcp/datastore/v1new/datastoreio.py]: done

**Comment by post-l** [comment]
After adding the class comments, I realized that we could go with removing the project attribute from the `Key` class. `WriteToDatastore` and `DeleteFromDatastore` would pass the  transform's project value  to the `to_client_entity` method on `Entity` and `Key`.

This would ease reading from a project and writing to another (removing the need to add a transform step to remove or update the project fields on the keys).

I guess this could be discuss in a new issue.

**Pull request merged**